File: cpp/sneaky_stack_alloc.cpp
#include <iostream>
using namespace std;

template<typename T> T id(T x) { return x; }

int *mk_int() { int x; return id(&x); }

int *sum(int lo, int hi)
{
  int *acc = mk_int();
  for(int i = lo; i <= hi; ++i)
    *acc += i;

  return acc;
}

int main(int, char **)
{
  for(int i = 0; i < 1000; ++i) {
    int *x = sum(0, i);
    cout << "Sum from 0 to " << i << ": " << *x << "\n";
  }

  return 0;
}

File: rs/src/sneaky_stack_alloc.rs
#[cfg(feature = "never")]
mod will_fail {
    fn id<T>(x : T) -> T { x }

    fn mk_int<'a>() -> &'a mut i32 { let mut x : i32; id(&mut x) }

    fn sum<'a>(lo : i32, hi : i32) -> &'a i32 {
        let acc = mk_int();
        for i in 0..(hi+1) { *acc += i };
        acc
    }
}
